### PR TITLE
IGNITE-14555 Create table from query result

### PR DESCRIPTION
**Pull request by alex-plekhanov**
Thank you for submitting the pull request to the Apache Ignite.

In order to streamline the review of the contribution 
we ask you to ensure the following steps have been taken:

### The Contribution Checklist
- [ ] There is a single JIRA ticket related to the pull request. 
- [ ] The web-link to the pull request is attached to the JIRA ticket.
- [ ] The JIRA ticket has the _Patch Available_ state.
- [ ] The pull request body describes changes that have been made. 
The description explains _WHAT_ and _WHY_ was made instead of _HOW_.
- [ ] The pull request title is treated as the final commit message. 
The following pattern must be used: `IGNITE-XXXX Change summary` where `XXXX` - number of JIRA issue.
- [ ] A reviewer has been mentioned through the JIRA comments 
(see [the Maintainers list](https://cwiki.apache.org/confluence/display/IGNITE/How+to+Contribute#HowtoContribute-ReviewProcessandMaintainers)) 
- [ ] The pull request has been checked by the Teamcity Bot and 
the `green visa` attached to the JIRA ticket (see [TC.Bot: Check PR](https://mtcga.gridgain.com/prs.html))

### Notes
- [How to Contribute](https://cwiki.apache.org/confluence/display/IGNITE/How+to+Contribute)
- [Coding abbreviation rules](https://cwiki.apache.org/confluence/display/IGNITE/Abbreviation+Rules)
- [Coding Guidelines](https://cwiki.apache.org/confluence/display/IGNITE/Coding+Guidelines)
- [Apache Ignite Teamcity Bot](https://cwiki.apache.org/confluence/display/IGNITE/Apache+Ignite+Teamcity+Bot)

If you need any help, please email dev@ignite.apache.org or ask anу advice on http://asf.slack.com _#ignite_ channel.

### CODE REVIEW

File: modules/calcite/src/main/codegen/includes/parserImpls.ftl [hunk 1]
```diff
@@ -148,19 +148,29 @@ SqlCreate SqlCreateTable(Span s, boolean replace) :
...
+        <AS> query = OrderedQueryOrExpr(ExprContext.ACCEPT_QUERY)
+    |
+        { query = null; }
+    )
     (
         <WITH> { s.add(this); } optionList = CreateTableOptionList()
     |
         { optionList = null; }
     )
     {
```
Comment (korlov42) [inline modules/calcite/src/main/codegen/includes/parserImpls.ftl:172]: Please extend `org.apache.ignite.internal.processors.query.calcite.sql.SqlDdlParserTest` with tests to verify the new clause format

File: modules/calcite/src/main/codegen/includes/parserImpls.ftl [hunk 2]
```diff
@@ -148,19 +148,29 @@ SqlCreate SqlCreateTable(Span s, boolean replace) :
...
     final SqlNodeList optionList;
+    final SqlNode query;
 }
 {
     <TABLE>
     ifNotExists = IfNotExistsOpt()
     id = CompoundIdentifier()
-    columnList = TableElementList()
+    (
+        columnList = TableElementList()
```
[the remark this answers is not in the document]
Comment (alex-plekhanov) [reply modules/calcite/src/main/codegen/includes/parserImpls.ftl:184]: You are right, thanks, fixed

File: modules/calcite/src/main/codegen/includes/parserImpls.ftl [hunk 3]
```diff
@@ -148,19 +148,29 @@ SqlCreate SqlCreateTable(Span s, boolean replace) :
...
+    |
+        { columnList = null; }
+    )
+    (
+        <AS> query = OrderedQueryOrExpr(ExprContext.ACCEPT_QUERY)
```
[the remark this answers is not in the document]
Comment (alex-plekhanov) [reply modules/calcite/src/main/codegen/includes/parserImpls.ftl]: Fixed

File: modules/calcite/src/main/codegen/includes/parserImpls.ftl (hunk 1 above)
Comment (alex-plekhanov) [reply modules/calcite/src/main/codegen/includes/parserImpls.ftl:196]: New test added

File: modules/calcite/src/main/java/org/apache/ignite/internal/processors/query/calcite/exec/PlannerHelper.java [hunk 4]
```diff
@@ -107,6 +121,72 @@ public static IgniteRel optimize(SqlNode sqlNode, IgnitePlanner planner, IgniteL
         }
     }
 
+    /**
+     * Creates physical plan for "INSERT INTO table SELECT ..." based on "CREATE TABLE table AS SELECT ..." statement.
+     */
+    public static IgniteRel makeCreateTableAsSelectPlan(CreateTableCommand createTableCmd, PlanningContext ctx,
```
[the remark this answers is not in the document]
Comment (alex-plekhanov) [reply modules/calcite/src/main/java/org/apache/ignite/internal/processors/query/calcite/exec/PlannerHelper.java]: > Have you considered an implicit rewriting of query with AS clause to two distinct statements (CREATE + INSERT)?
I've changed it to CREATE+INSERT, but now there is yet another planning phase on the execution phase for CREATE AS SELECT statement. Changes ready for review, please have a look again.

File: modules/calcite/src/test/java/org/apache/ignite/internal/processors/query/calcite/integration/TableDdlIntegrationTest.java [hunk 5]
```diff
@@ -275,6 +275,97 @@ public void createTableOnExistingCache() {
...
+
+        assertEquals(1, row.get(0));
+        assertEquals("test", row.get(1));
+    }
+
+    /**
+     * Creates table with specified columns as select.
+     */
+    @Test
+    public void createTableAsSelectWithColumns() {
```
Comment (korlov42) [inline modules/calcite/src/test/java/org/apache/ignite/internal/processors/query/calcite/integration/TableDdlIntegrationTest.java:298]: let's add a few test to verify column names of just created tables

**Review by korlov42 — APPROVED** [review]

File: modules/calcite/src/main/codegen/includes/parserImpls.ftl [hunk 6]
```diff
@@ -142,25 +146,55 @@ SqlNodeList TableElementList() :
...
+SqlNodeList TableSimpleColumnList() :
+{
+    final Span s;
+    final List<SqlNode> list = new ArrayList<SqlNode>();
+    SqlIdentifier id;
+}
+{
+    <LPAREN> { s = span(); }
+    id = SimpleIdentifier() {
+        list.add(SqlDdlNodes.column(s.add(id).end(this), id, null, null, null));
```
[the remark this answers is not in the document]
Comment (alex-plekhanov) [reply modules/calcite/src/main/codegen/includes/parserImpls.ftl]: Ok, fixed

File: modules/calcite/src/test/java/org/apache/ignite/internal/processors/query/calcite/integration/TableDdlIntegrationTest.java (hunk 5 above)
Comment (alex-plekhanov) [reply modules/calcite/src/test/java/org/apache/ignite/internal/processors/query/calcite/integration/TableDdlIntegrationTest.java:298]: Check for column names added to this test

**Comment by alex-plekhanov** [comment]
Merged to sql-calcite branch

**Pull request closed without merging**